snapcraft/internal/pluginhandler.py

```diff
@@ -650,15 +655,46 @@ def _migratable_filesets(fileset, srcdir):
     return snap_files, snap_dirs
 
 
-def _migrate_files(snap_files, snap_dirs, srcdir, dstdir, missing_ok=False,
-                   follow_symlinks=False, fixup_func=lambda *args: None):
+def _migrate_files_directories(snap_dirs, srcdir, dstdir,
+                               follow_symlinks=False, keep_owner=False):
     for directory in snap_dirs:
-        os.makedirs(os.path.join(dstdir, directory), exist_ok=True)
+        src = os.path.join(srcdir, directory)
+        dst = os.path.join(dstdir, directory)
+        os.makedirs(dst, exist_ok=True)
+        if keep_owner:
+            try:
+                st = os.stat(src, follow_symlinks=follow_symlinks)
+                uid = st.st_uid
+                gid = st.st_gid
+                os.chown(dst, uid, gid, follow_symlinks=follow_symlinks)
```

Comment:
There are build steps that set a few file ownerships such as /home/ubuntu

On Thu, Jul 28, 2016 at 3:07 PM, Kyle Fazzari <notifications@github.com>
wrote:

> In snapcraft/internal/pluginhandler.py
> <https://github.com/snapcore/snapcraft/pull/690#discussion_r72683018>:
>
> >      for directory in snap_dirs:
> > -        os.makedirs(os.path.join(dstdir, directory), exist_ok=True)
> > +        src = os.path.join(srcdir, directory)
> > +        dst = os.path.join(dstdir, directory)
> > +        os.makedirs(dst, exist_ok=True)
> > +        if keep_owner:
> > +            try:
> > +                st = os.stat(src, follow_symlinks=follow_symlinks)
> > +                uid = st.st_uid
> > +                gid = st.st_gid
> > +                os.chown(dst, uid, gid, follow_symlinks=follow_symlinks)
>
> I know ogra said this solved the issue, but I don't quite understand how.
> The LP builders run as root, which means the files are owned by root. How
> are uid and gid not root here?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/snapcore/snapcraft/pull/690/files/2e9fb407545a200a22453495a82c2c07b29aeb47#r72683018>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnbyNjEc2gEYTxqjTdFOpQqh7pPT8Ulks5qaP3UgaJpZM4JWK_l>
> .
>